### PR TITLE
9512 imperfection

### PR DESCRIPTION
**Pull request by david-liska**
# Description

Please include a summary of the change and which issue is fixed. Please also include relevant motivation and context. List any dependencies that are required for this change.

Fixes # (issue)

## Type of change

Please delete options that are not relevant.

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

# How Has This Been Tested?

Please describe the tests that you ran to verify your changes. Provide instructions so we can reproduce. Please also list any relevant details for your test configuration

- [ ] Unit Tests
- [ ] Attached examples
- [ ] Own example attached to Pull Request

**Test Configuration**:
* RFEM / RSTAB / RSECTION version:
# Checklist:

- [ ] My code follows the style guidelines of this project
- [ ] I have performed a self-review of my own code
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [ ] My changes generate no new warnings
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] New and existing unit tests pass locally with my changes
- [ ] Any dependent changes have been merged and published in downstream modules

### CODE REVIEW

**Review by jarabroz — CHANGES_REQUESTED** [review]

File: examples/high_level_functions_demos/imperfections_demo.js [hunk 1]
```diff
@@ -0,0 +1,194 @@
...
+var nodesForMembers = createNodesGrid(1, 8, [8, 1], [3, 0]);
+
+var member2 = new Member;
+var members_list = [];
+
+for (var i = 0; i < nodesForMembers.length - 1; ++i) {
+    members_list.push(member2.Beam(undefined, [nodesForMembers[i], nodesForMembers[i + 1]], section.GetNo()));
+}
+
+var memberSet = new MemberSet;
```
Comment (jarabroz) [inline examples/high_level_functions_demos/imperfections_demo.js:43]: Missing ()

File: examples/high_level_functions_demos/imperfections_demo.js [hunk 2]
```diff
@@ -0,0 +1,194 @@
...
+memberSet.ContinuousMembers(undefined, members_list);
+
+if (RFEM) {
+    var thickness = new Thickness;
```
Comment (jarabroz) [inline examples/high_level_functions_demos/imperfections_demo.js:47]: Missing ()

File: examples/high_level_functions_demos/imperfections_demo.js [hunk 3]
```diff
@@ -0,0 +1,194 @@
...
+    thickness.Uniform(1, "Thickness", material.GetNo(), 0.250)
```
Comment (jarabroz) [inline examples/high_level_functions_demos/imperfections_demo.js:48]: Missing semicolon

File: examples/high_level_functions_demos/imperfections_demo.js [hunk 4]
```diff
@@ -0,0 +1,194 @@
...
+
+    var nodesForSurfaces = createNodesGrid(1, 10, [6, 2], [3, 3]);
+    dictSurfaces = createSurfacesFromNodesGrid(nodesForSurfaces, [3, 1], surfaces.TYPE_STANDARD, thickness.GetNo());
+    var surfaceList = Object.keys(dictSurfaces).map(function(key){
+        return dictSurfaces[key][0];
+    });
+
+    var surfaceSet = new SurfaceSet;
```
Comment (jarabroz) [inline examples/high_level_functions_demos/imperfections_demo.js:56]: Missing ()

File: examples/high_level_functions_demos/imperfections_demo.js [hunk 5]
```diff
@@ -0,0 +1,194 @@
...
+member_set_imperfection.InitialBowAndCriterion(undefined, localIC, [memberSet.GetNo()], "PRINCIPAL", "V_NEGATIVE");
+member_set_imperfection.Relative(1.5, "DIN_18800");
+member_set_imperfection.InitialBowAndCriterion(undefined, localIC, [memberSet.GetNo()], "PRINCIPAL", "V_NEGATIVE");
+member_set_imperfection.Relative(1.5, "DEFINE", 2.5);
+member_set_imperfection.InitialBowAndCriterion(undefined, localIC, [memberSet.GetNo()], "PRINCIPAL", "V_NEGATIVE", true);  // With reference to list of members
+member_set_imperfection.Absolute(0.5, "EN_1993");
+
+if (RFEM) {
+    /****************************************************** Surface imperfections *************************************************************/
+    var surfaceImperfection = new SurfaceImperfection;
```
Comment (jarabroz) [inline examples/high_level_functions_demos/imperfections_demo.js:182]: Missing ()

File: examples/high_level_functions_demos/imperfections_demo.js [hunk 6]
```diff
@@ -0,0 +1,194 @@
...
+    surfaceImperfection.Relative(undefined, localIC, [surfaceList[0].no], 2, 300, "LOCAL_Z_NEGATIVE");
+    surfaceImperfection.Absolute(undefined, localIC, [surfaceList[1].no], 3, "LOCAL_Z_NEGATIVE");
+
+    /****************************************************** Surface set imperfections *********************************************************/
+    var surfaceSetImperfection = new SurfaceSetImperfection;
```
Comment (jarabroz) [inline examples/high_level_functions_demos/imperfections_demo.js:187]: Missing ()

File: includes/Imperfections/ImperfectionSupport.js [hunk 7]
```diff
@@ -0,0 +1,275 @@
+function GetMemberImperfectionDirection(coordinate_system, imperfection_direction) {
+    if (coordinate_system === "LOCAL") {
+        const imperfection_directions_dict = {
+            "LOCAL_Y" : member_imperfections.IMPERFECTION_DIRECTION_LOCAL_Y,
+            "LOCAL_Z" : member_imperfections.IMPERFECTION_DIRECTION_LOCAL_Z,
+            "LOCAL_Y_NEGATIVE" : member_imperfections.IMPERFECTION_DIRECTION_LOCAL_Y_NEGATIVE,
+            "LOCAL_Z_NEGATIVE" : member_imperfections.IMPERFECTION_DIRECTION_LOCAL_Z_NEGATIVE
+        }
```
Comment (jarabroz) [inline includes/Imperfections/ImperfectionSupport.js:8]: Missing semicolon

File: includes/BasicObjects/MemberSet.js [hunk 8]
```diff
@@ -22,6 +22,20 @@ function MemberSet(no,
...
+ * @returns Number of member set
+ */
+MemberSet.prototype.GetNo = function() {
+    return this.member_set.no;
+};
+
+/**
+ * @returns Member set object
+ */
+MemberSet.prototype.MemberSet = function() {
```
Comment (jarabroz) [inline includes/BasicObjects/MemberSet.js:35]: GetMemberSet

File: examples/high_level_functions_demos/imperfections_demo.js [hunk 9]
```diff
@@ -0,0 +1,194 @@
...
+imperfection_case.StaticDeformation(undefined, [lc6.GetNo()], "LOAD_CASE", lc7.GetNo(), 1.5, "SPECIFIC_NODE", node.no, "USER_DEFINED_IN_V");
+imperfection_case.GroupOfImperfection(undefined, [lc7.GetNo()], [[lc3.GetNo(), 3.0, "Comment 1"]], false);
+var coordinate_system = new CoordinateSystem();
+coordinate_system.Offset(undefined, [1.0, 2.0, 3.0]);
+
+/****************************************************** Member imperfections *************************************************************/
+var member_imperfection = new MemberImperfection();
+member_imperfection.InitialSway(undefined, localIC, [member.GetNo()], "PRINCIPAL", "V_NEGATIVE");
+member_imperfection.Relative(1.5);
+member_imperfection.InitialSway(undefined, localIC, [member.GetNo()]);
```
Comment (jarabroz) [inline examples/high_level_functions_demos/imperfections_demo.js:82]: Make a new variable for this -> it is hard to read. I spent some time to decript it how it works. So I create MemberImperfection object then I use function for selection of type of initial sway.

**Review by jarabroz — CHANGES_REQUESTED** [review]
Please add imperfection case and local imperfection to steel hall example. There is the old way to create them.

File: includes/Imperfections/ImperfectionCase.js [hunk 10]
```diff
@@ -0,0 +1,484 @@
...
+        this.imperfectionCase.direction_for_level_direction = GetImperfectionDirection(this.imperfectionCase.coordinate_system.type, imperfection_direction);
+    }
+    if (typeof sway_coefficients_reciprocal === "undefined") {
+        sway_coefficients_reciprocal = true;
+    }
+    this.imperfectionCase.sway_coefficients_reciprocal = sway_coefficients_reciprocal;
+    ASSERT(Array.isArray(level_imperfections) && level_imperfections.length >= 1, "Level imperfections must be array of arrays (first item contains only level value, the other contain level, theta_x, theta_y");
+    for (var i = 0; i < level_imperfections.length; ++i) {
+        var row = this.imperfectionCase.level_imperfections.row_count();
+        this.imperfectionCase.level_imperfections.insert_row(row);
```
[the remark this answers is not in the document]
Comment (david-liska) [reply includes/Imperfections/ImperfectionCase.js:167]: There are two way how add and edit rows in table. In this case we get index of new row (row_count()) and then we insert this new row to edit it.

**Review by jarabroz — APPROVED** [review]

**Pull request merged**